src/ripple/app/misc/impl/AmendmentTable.cpp

```diff
@@ -539,12 +554,36 @@ AmendmentTableImpl::needValidatedLedger (LedgerIndex ledgerSeq)
 }
 
 void
-AmendmentTableImpl::doValidatedLedger (
+AmendmentTableImpl::doValidatedLedger(
     LedgerIndex ledgerSeq,
-    std::set<uint256> const& enabled)
+    std::set<uint256> const& enabled,
+    majorityAmendments_t const& majority)
 {
     for (auto& e : enabled)
         enable(e);
+
+    std::lock_guard sl(mutex_);
+    // Since we have the whole list in `majority`, reset the time flag, even if
+    // it's currently set. If it's not set when the loop is done, then any
+    // prior unknown amendments have lost majority.
+    firstUnsupportedExpected_.reset();
+    for (auto& m : majority)
+    {
+        auto s = add(m.first);
+
+        if (s->enabled)
+            break;
+
+        if (!s->supported)
+        {
+            JLOG(j_.info()) << "Unsupported amendment " << m.first
+                             << " reached majority at " << to_string(m.second);
+            if (!firstUnsupportedExpected_ || firstUnsupportedExpected_ > m.second)
+                firstUnsupportedExpected_ = m.second;
```

Comment:
It's a pity that the [semantics of relational operators for `boost::optional`](https://www.boost.org/doc/libs/1_72_0/libs/optional/doc/html/boost_optional/tutorial/relational_operators.html) are the reverse of what we want here, so they won't allow us to simplify this:

>In a similar manner, type `optional<T>` is `LessThanComparable` whenever `T` is `LessThanComparable`. The optional object containing no value is compared less than any value of `T`.

Reply:
Agreed, though I suppose one way to look at it is that they (presumably) have to put unassigned at one end of the value space or the other. If I was looking for the largest, it would have been just fine.